nni/retiarii/oneshot/pytorch/differentiable.py

```diff
@@ -0,0 +1,325 @@
+        detached_x, output = ctx.saved_tensors
+
+        grad_x = torch.autograd.grad(output, detached_x, grad_output, only_inputs=True)
+        # compute gradients w.r.t. binary_gates
+        binary_grads = ctx.backward_func(detached_x.data, output.data, grad_output.data)
+
+        return grad_x[0], binary_grads, None, None
+
+
+class ProxylessLayerChoice(nn.Module):
+    def __init__(self, ops):
+        super(ProxylessLayerChoice, self).__init__()
+        self.ops = nn.ModuleList(ops)
+        self.alpha = nn.Parameter(torch.randn(len(self.ops)) * 1E-3)
+        self._binary_gates = nn.Parameter(torch.randn(len(self.ops)) * 1E-3)
+        self.sampled = None
+
+    def forward(self, *args, **kwargs):
+        if self.training:
+            def run_function(ops, active_id, **kwargs):
+                def forward(_x):
+                    return ops[active_id](_x, **kwargs)
+                return forward
+
+            def backward_function(ops, active_id, binary_gates, **kwargs):
+                def backward(_x, _output, grad_output):
+                    binary_grads = torch.zeros_like(binary_gates.data)
+                    with torch.no_grad():
+                        for k in range(len(ops)):
+                            if k != active_id:
+                                out_k = ops[k](_x.data, **kwargs)
+                            else:
+                                out_k = _output.data
+                            grad_k = torch.sum(out_k * grad_output)
+                            binary_grads[k] = grad_k
+                    return binary_grads
+                return backward
+
+            assert len(args) == 1
+            x = args[0]
+            return _ArchGradientFunction.apply(
+                x, self._binary_gates, run_function(self.ops, self.sampled, **kwargs),
+                backward_function(self.ops, self.sampled, self._binary_gates, **kwargs)
+            )
+
+        return super().forward(*args, **kwargs)
+
+    def resample(self):
+        probs = F.softmax(self.alpha, dim=-1)
+        sample = torch.multinomial(probs, 1)[0].item()
+        self.sampled = sample
+        with torch.no_grad():
+            self._binary_gates.zero_()
+            self._binary_gates.grad = torch.zeros_like(self._binary_gates.data)
+            self._binary_gates.data[sample] = 1.0
+
+    def finalize_grad(self):
+        binary_grads = self._binary_gates.grad
+        with torch.no_grad():
+            if self.alpha.grad is None:
+                self.alpha.grad = torch.zeros_like(self.alpha.data)
+            probs = F.softmax(self.alpha, dim=-1)
+            for i in range(len(self.ops)):
+                for j in range(len(self.ops)):
+                    self.alpha.grad[i] += binary_grads[j] * probs[j] * (int(i == j) - probs[i])
+
+    def export(self):
+        return torch.argmax(self.alpha).item()
+
+    def export_prob(self):
+        return F.softmax(self.alpha, dim=-1)
+
+
+class ProxylessInputChoice(nn.Module):
```

Comment:
In previous version of ProxylessNAS, input choice is not supported. Why support InputChoice here? Maybe some docstring could help understanding.

Reply:
The way of two ops are similar, and the inputchoice was not supported due to some previous limitation.